src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -392,6 +392,61 @@ public boolean delete(Object pathObject, StarlarkThread thread)
     }
   }
 
+  @StarlarkMethod(
+      name = "rename",
+      doc = "Renames the file or directory from `src` to `dst`.",
+      useStarlarkThread = true,
+      parameters = {
+        @Param(
+            name = "src",
+            allowedTypes = {
+              @ParamType(type = String.class),
+              @ParamType(type = Label.class),
+              @ParamType(type = StarlarkPath.class)
+            },
+            doc = "The path of the existing file or directory to rename."),
+        @Param(
+            name = "dst",
+            allowedTypes = {
+              @ParamType(type = String.class),
+              @ParamType(type = Label.class),
+              @ParamType(type = StarlarkPath.class)
+            },
+            doc = "The new name of the file or directory."),
+      })
+  public void rename(Object srcName, Object dstName, StarlarkThread thread)
+      throws RepositoryFunctionException, EvalException, InterruptedException {
+    StarlarkPath srcPath = getPath(srcName);
+    StarlarkPath dstPath = getPath(dstName);
+    WorkspaceRuleEvent w =
+        WorkspaceRuleEvent.newRenameEvent(
+            srcPath.toString(),
+            dstPath.toString(),
+            identifyingStringForLogging,
+            thread.getCallerLocation());
+    env.getListener().post(w);
+    try {
+      checkInOutputDirectory("write", dstPath);
+      makeDirectories(dstPath.getPath());
+      srcPath.getPath().renameTo(dstPath.getPath());
```

Comment:
This rename may fail if src and dst are on different filesystems, which is not something you could easily detect in a repo rule. How about using `FileSystemUtils.moveFile` instead, which will fall back to copy instead?

Reply:
I think it's fine to fail if the source and destination are on separate filesystems, since that's something that would be really difficult to have happen by accident.

To clarify, `rename()` is for two paths within the same repository workspace being set up by the repository rule. It's not renaming across the workspace directory boundary.

Reply:
That makes sense. In that case, could you enforce that `src` also lies in the output directory?

Reply:
Yep, added that in the most recent snapshot (? or whatever GitHub calls it):
```java
      checkInOutputDirectory("write", srcPath);
      checkInOutputDirectory("write", dstPath);
```